### DISCUSSION TITLE
Changing shape size leads to issue

### DISCUSSION THREAD

**Discussion by KlassenKonstantin** [dyn4j/dyn4j#239] [Q&A]
Hello! Thanks for the lib! Let's start directly with a video:

https://user-images.githubusercontent.com/1836066/200139210-e905b624-079d-4376-bb68-4ea108371b79.mov

So basically everything works super fine, but when I start to change the size of a shape, by removing and adding adapted fixtures, it looks like in the video. What can I do to fix that? I can share more info if the video is not enough :)

Thanks in advance!

EDIT: increasing `positionConstraintSolverIterations` and `velocityConstraintSolverIterations` does not help

**Comment by wnbittle** [comment]
Anyway you can share your code?  

The only thing I can think of is make sure you are calling the `updateMass` method after adding/removing fixtures.

Another thing to try is to remove the body entirely and just add a new one at the same position and see if you get the same behavior.

**Reply by KlassenKonstantin** [reply]
No success :/. `updateMass` was already included in the video and swapping bodies instead of fixtures also did not help

Stack is Android / Compose / Kotlin / Coroutines

```kotlin
class Simulation internal constructor(
    val scale: Double,
    private val world: World<Body>
) {
    internal val renderBodies = mutableStateMapOf<String, RenderBody>()

    suspend fun run() {
        var last = System.currentTimeMillis()
        while (true) {
            val now = System.currentTimeMillis()
            val elapsed = (now - last).toDouble() / 1000.0
            last = now

            world.update(elapsed)
            world.bodies.forEach { body ->
                (body.userData as? BodyMetaData)?.let { metaData ->
                    renderBodies[metaData.id] = body.toRenderBody(scale)
                }
            }
            delay(10)
        }
    }

    internal fun updateWorldSize(intSize: IntSize) {
        world.updateWorldSize(
            intSize.width / scale,
            intSize.height / scale,
        )
    }

    internal fun removeBodiesWithIdNotIn(currentIds: List<String>) {
        val removedBodyIds = world.removeBodiesWithIdNotIn(currentIds)
        removedBodyIds.forEach {
            renderBodies.remove(it)
        }
    }

    internal fun updateBodyMetaData(newBodyMetaData: BodyMetaData) {
        val body = world.bodies.find { (it.userData as? BodyMetaData)?.id == newBodyMetaData.id } ?: return
        val currentMetaData = body.userData as BodyMetaData
        if (currentMetaData == newBodyMetaData) return

        body.userData = newBodyMetaData
        if (newBodyMetaData.height > 0 && newBodyMetaData.width > 0) {
            body.removeAllFixtures()
            body.addFixture(createFixture(newBodyMetaData), 1.0, 0.2, 0.4)
            body.updateMass()
        }
    }

    internal fun createBody(bodyMetaData: BodyMetaData) {
        world.addBody(createBodyInternal(bodyMetaData))
    }

    private fun createBodyInternal(bodyMetaData: BodyMetaData) = Body().also { newBody ->
        newBody.isAtRestDetectionEnabled = false
        newBody.userData = bodyMetaData
        newBody.addFixture(createFixture(bodyMetaData), 1.0, 0.2, 0.4)
        newBody.setMass(MassType.NORMAL)
    }
    private fun createFixture(bodyMetaData: BodyMetaData): Convex = when {
        bodyMetaData.shape.isCircle() -> Geometry.createCircle(bodyMetaData.height / 2)
        else -> Geometry.createRectangle(bodyMetaData.width, bodyMetaData.height)
    }
}
```

**Comment by wnbittle** [comment]
Can you provide a dump of the simulations initial configuration?   Shapes, sizes, mass, position, rotation, etc?

**Comment by KlassenKonstantin** [comment]
It run very smooth, when I reduce the delay between the loops to 1 and use `world.updatev`. Why can't I have the same calculations when I use a delay of 100? I thought that is what the elapsed time is for 🤔

**Reply by wnbittle** [reply]
Ahh, I see.  `world.updatev` will update the world _every_ time it's called.  All the other `update` methods will only update the simulation if the step frequency has elapsed (by default 1/60 seconds - 60 times a second).

My guess is that it's smooth when you use `updatev` and a delay of 1 because you are now running the simulation as fast as possible.
